### ISSUE TITLE
fix_random_seed problem with devices

### ISSUE THREAD

**Issue by danpovey** [lhotse-speech/lhotse#1564]
I believe torch.manual_seed(), called from lhotse.fix_random_seed() affects seeds on *all devices*, i.e. all visible GPUs, not just ones this process is creating models on.  I think there could be a problem in some circumstances when fix_random_seed() gets called from things like data-loader worker processes, that could lead to nondeterminism.

https://github.com/lhotse-speech/lhotse/blob/c72136cdca885a88f05e0ca62b920600980e859b/lhotse/utils.py#L141

We have some code like this:
```
class _SeedWorkers:
    def __init__(self, seed: int):
        self.seed = seed

    def __call__(self, worker_id: int):
        fix_random_seed(self.seed + worker_id)
```

```
       worker_init_fn = _SeedWorkers(seed)
        train_dl = DataLoader(
            train,
            sampler=train_sampler,
            batch_size=None,
            num_workers=self.args.num_workers,
            persistent_workers=False,
            worker_init_fn=worker_init_fn,
        )
```
and I think this might be a cause of nondeterminism on CUDA devices.  What happens, I think, is that fix_random_seed() affects the seed of all visible devices, and here it's probably arbitrary which worker-id gets called last.   In any case, without torch.distributed.barrier() getting called before and after these types of call, we don't really know *when* the seed is getting fixed, which could affect the random state used by stragglers and early-starters.

I have noticed some nondeterminism appearing in my training (i.e. it's not the same when I re-run), that I believe was coming from the CUDA torch seed; and certainly not the seed used in random.random(), when I changed some code to initialize the dataloaders once per epoch.   I'm going to try re-setting the random seed before and after torch.distributed.barrier() right after doing this, to make absolutely sure the seed is deterministic.  I just wondered whether you (Piotr) have any thoughts on the right way to deal with this

**Comment by pzelasko** [comment]
Hey Dan. Sorry I missed your post earlier. 

I tried to test this hypothesis with the following snippet:

```python

 # seed_test.py
 # Run with: torchrun --nproc_per_node=2 seed_test.py

 import os
 import torch
 import torch.distributed as dist

 def main():
     dist.init_process_group(backend="nccl")
     local_rank = int(os.environ["LOCAL_RANK"])
     torch.cuda.set_device(local_rank)
     device = torch.device(f"cuda:{local_rank}")

     num_visible = torch.cuda.device_count()
     print(f"[rank {local_rank}] Can see {num_visible} GPU(s). "
           f"Active device: cuda:{torch.cuda.current_device()}\n")

     dist.barrier()

     SEED = 42
     if dist.get_rank() == 0:
         torch.random.manual_seed(SEED)

     dist.barrier()

     weight = torch.randn(3, 3, device=device)
     print(f"[rank {local_rank}] Same seed={SEED} → weight[0]: {weight[0].tolist()}\n")

     dist.barrier()

 if __name__ == "__main__":
     main()
```

which outputs

```
[rank 1] Can see 2 GPU(s). Active device: cuda:1
[rank 0] Can see 2 GPU(s). Active device: cuda:0
[rank 0] Same seed=42 → weight[0]: [0.19401879608631134, 2.1613736152648926, -0.17205022275447845]
[rank 1] Same seed=42 → weight[0]: [-0.25486916303634644, -1.5415321588516235, -0.9476635456085205]
```

Both GPUs are visible to each process. If `torch.random.manual_seed` had seeded both of them, then we should see the same randn result - but GPU 1 has non-deterministic output across invocations unlike GPU 0. 
Even removing `torch.cuda.set_device(local_rank)` did not change the outcome.

That said `fix_random_seed` probably doesn't cover every imaginable scenario... for 100% determinism it could be insufficient depending on your setup (eg if you use model parallelism techniques like TP you might need to manage context specific RNGs).

**Comment by pzelasko** [comment]
Oh, I just noticed this bit:

```python
class _SeedWorkers:
    def __init__(self, seed: int):
        self.seed = seed

    def __call__(self, worker_id: int):
        fix_random_seed(self.seed + worker_id)
```

I think you have a race condition with multiple workers seeding the same GPU. 
That's indeed undesirable. Maybe it's better if lhotse doesn't touch GPU RNG at all and only seeds CPU devices?

**Comment by danpovey** [comment]
You did: `torch.random.manual_seed(SEED)`, I can't see docs for this online, shouldn't it be `torch.cuda.manual_seed(SEED)`?

**Comment by danpovey** [comment]
Got this from google:

In PyTorch, the random state is specific to the process and also device-specific (CPU vs. GPU).While functions like torch.manual_seed(seed) set the seed for both CPU and all GPU devices, they maintain separate random number generator (RNG) states.

So since the worker processes are separate, they do not set each others' random seeds.  In SeedWorkers, I'm assuming these are actually separate worker processes also, since just being threads wouldn't make sense for the purpose (given python's GIL).  So they wouldn't be setting a random state for the whole device.  In fact I see now that it would be impossible for the random state to be global across all processes.

**Comment by pzelasko** [comment]
Yes I believe you are correct. I’m still thinking it’d be best for Lhotse not to interact with GPU random state and let the users set it in their training loop script instead. I think I was writing this thinking that it would only seed CPU RNGs since it wasn’t the torch.cuda variant. I’ll make the PR. 

The docs for the function I used are here https://docs.pytorch.org/docs/2.11/random.html#torch.random.manual_seed

**Comment by shanguanma** [comment]
```
# seed_test2.py
# Run with: torchrun --nproc_per_node=2 seed_test2.py
import os
import torch
import torch.distributed as dist
 
 
def main():
    dist.init_process_group(backend="nccl")
    local_rank = int(os.environ["LOCAL_RANK"])
    torch.cuda.set_device(local_rank)
    device = torch.device(f"cuda:{local_rank}")
    num_visible = torch.cuda.device_count()
    print(
        f"[rank {local_rank}] Can see {num_visible} GPU(s). "
        f"Active device: cuda:{torch.cuda.current_device()}\n"
    )
    dist.barrier()
 
    SEED = 42
    #if dist.get_rank() == 0:
    #    torch.random.manual_seed(SEED)
    torch.cuda.manual_seed_all(SEED)
    dist.barrier()
 
    weight = torch.randn(3, 3, device=device)
    print(f"[rank {local_rank}] Same seed={SEED} → weight[0]: {weight[0].tolist()}\n")
    dist.barrier()
 
 
if __name__ == "__main__":
    main()
```
which outputs

```
[rank 0] Can see 2 GPU(s). Active device: cuda:0
[rank 1] Can see 2 GPU(s). Active device: cuda:1
[rank 1] Same seed=42 → weight[0]: [0.19401879608631134, 2.1613736152648926, -0.17205022275447845]
[rank 0] Same seed=42 → weight[0]: [0.19401879608631134, 2.1613736152648926, -0.17205022275447845]
```
